AppiumForMac/Server/Controller/AfMSessionController.m

```diff
@@ -1190,6 +1201,107 @@ - (NSArray *)findAllUsingAXPathComponents:(NSArray *)axPathComponents rootUIElem
     return @[];
 }
 
+-(PFUIElement*) findElementRecursively:(PFUIElement *)rootUIElement calls:(int)calls
+{
+    if ([self checkIfElementFound:rootUIElement])
+    {
+        return rootUIElement;
+    }
+    calls++;
+    
+    NSArray* children = rootUIElement.AXChildren;
+    
+    PFUIElement* result = nil;
+    
+    if (children)
+    {
+        for (PFUIElement* node in children)
+        {
+            result = [self findElementRecursively:node calls:calls];
+            if (result)
+            {
+                return result;
+            }
+        }
+    }
+    
+    calls--;
+    return nil;
+}
+
+-(NSArray*) findAllUsingPartialAXPath:(NSArray *)axPathComponents rootUIElement:(PFUIElement *)rootUIElement
+{
+    NSArray* childUIElements = rootUIElement.AXChildren;
+    
+    if (!childUIElements || [childUIElements count] == 0)
+    {
+        return @[];
+    }
+    
+    [self setRoleAndPredicatesOfElement:[axPathComponents lastObject]];
+    NSMutableArray *matchedChildren = [NSMutableArray array];
+    PFUIElement* foundElement = [self findElementRecursively:rootUIElement calls:0];
+    
+    if ([foundElement isNotEqualTo:nil])
+    {
+        [matchedChildren addObject:foundElement];
+    }
+    
+    return matchedChildren;
+}
+
+-(void) setRoleAndPredicatesOfElement:(NSString*) elementToFind
```

Comment:
This class seems overloaded with different stuff. Would it be possible to extract lookup-specific stuff into a separate class/category? Even better, add some basic unit tests to it?